File: src/VBNetFx/FacessoData/SQLServer/SPAccess_Users.vb
Imports Activedev
Imports System.Data
Imports System.Data.SqlClient
Imports System.Data.Common

Partial Public NotInheritable Class SPAccess

    ''' <summary>
    ''' Überprüft, ob der angegebene Benutzername für die Subsidiary schon existiert.
    ''' </summary>
    ''' <param name="IDSubsidiary">ID der verwendeten Subsidiary</param>
    ''' <param name="Username">Benutzername</param>
    ''' <param name="ExcludeIDUser">Benutzername</param>
    ''' <returns>True, wenn die Kostenstelle schon vorhanden war.</returns>
    ''' <remarks></remarks>
    Public Function Users_DoesUsernameExist(ByVal IDSubsidiary As Integer, _
                                                   ByVal Username As String, _
                                                   ByVal ExcludeIDUser As ADDBNullable(Of Integer)) _
                                                   As Boolean

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("Users_DoesUsernameExist", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = IDSubsidiary
                .Add("@Username", SqlDbType.nvarchar, 100).Value = Username
                .Add("@ExcludeIDUser", SqlDbType.Int).Value = ExcludeIDUser.Value
                .Add("@DoesExist", SqlDbType.Bit)
                .Item("@DoesExist").Direction = ParameterDirection.Output
            End With
            locCmd.ExecuteReader()
            Return CBool(locCmd.Parameters("@DoesExist").Value)
        End Using
    End Function

    ''' <summary>
    ''' Fügt den in UserInfo und AddressDetails gespeicherten Benutzer der Datenbank hinzu.
    ''' </summary>
    ''' <param name="ui">UserInfo mit den Hauptdaten für den Benutzer.</param>
    ''' <param name="addrDet">AddressDetailsInfo mit den Adressendaten für den Benutzer.</param>
    ''' <returns>Neue ID des Users.</returns>
    ''' <remarks>Die korrelierende Stored Procedure handelt die Versionskontrolle und schließt doppelte Benutzernamen aus.</remarks>
    Public Function Users_Add(ByVal ui As UserInfo, ByVal CreatedByIDUser As Integer, ByVal addrDet As AddressDetailsInfo) As Integer

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("Users_Add", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = ui.IDSubsidiary
                .Add("@IDCostCenter", SqlDbType.Int).Value = ui.IDCostCenter
                .Add("@FirstName", SqlDbType.NVarChar, 100).Value = ui.FirstName
                .Add("@LastName", SqlDbType.NVarChar, 100).Value = ui.LastName
                .Add("@UserName", SqlDbType.NVarChar, 100).Value = ui.Username
                .Add("@Password", SqlDbType.VarBinary, 128).Value = ui.Password
                .Add("@ClearanceLevel", SqlDbType.BigInt).Value = ui.ClearanceLevel
                .Add("@HasWorkstationAccess", SqlDbType.Bit).Value = ui.HasWorkstationAccess
                .Add("@HasInternetAccess", SqlDbType.Bit).Value = ui.HasInternetAccess
                .Add("@IsActivated", SqlDbType.Bit).Value = ui.IsActivated
                .Add("@DoesExpire", SqlDbType.Bit).Value = ui.DoesExpire
                .Add("@ExpireDate", SqlDbType.DateTime).Value = ui.ExpireDate
                .Add("@WasCurrentTo", SqlDbType.DateTime).Value = FacessoGeneric.OpenCurrentToDate
                .Add("@CreatedByIDUser", SqlDbType.Int).Value = CreatedByIDUser
                .Add("@Comment", SqlDbType.NText).Value = ui.Comment.Value

                'Adressen-Details
                .Add("@PersonnelNo", SqlDbType.Int).Value = addrDet.PersonnelNo.Value
                .Add("@MiddleName", SqlDbType.NVarChar, 100).Value = addrDet.MiddleName.Value
                .Add("@Title", SqlDbType.NVarChar, 100).Value = addrDet.Titel.Value
                .Add("@Street", SqlDbType.NVarChar, 100).Value = addrDet.Street.Value
                .Add("@Zip", SqlDbType.NVarChar, 10).Value = addrDet.Zip.Value
                .Add("@City", SqlDbType.NVarChar, 100).Value = addrDet.City.Value
                .Add("@CountryCode", SqlDbType.NVarChar, 10).Value = addrDet.CountryCode.Value
                .Add("@Country", SqlDbType.NVarChar, 100).Value = addrDet.Country.Value
                .Add("@CompanyPhone", SqlDbType.NVarChar, 100).Value = addrDet.CompanyPhone.Value
                .Add("@PrivatePhone", SqlDbType.NVarChar, 100).Value = addrDet.PrivatePhone.Value
                .Add("@CompanyEmail", SqlDbType.NVarChar, 255).Value = addrDet.CompanyEmail.Value
                .Add("@PrivateEmail", SqlDbType.NVarChar, 255).Value = addrDet.PrivateEmail.Value
                .Add("@CompanyMobile", SqlDbType.NVarChar, 100).Value = addrDet.CompanyMobile.Value
                .Add("@PrivateMobile", SqlDbType.NVarChar, 100).Value = addrDet.PrivateMobile.Value
                .Add("@URL", SqlDbType.NVarChar, 100).Value = addrDet.URL.Value
                .Add("@IDUserNew", SqlDbType.Int)
                .Item("@IDUserNew").Direction = ParameterDirection.Output
            End With
            locCmd.ExecuteReader()
            Return CInt(locCmd.Parameters("@IDUserNew").Value)
        End Using

    End Function

    ''' <summary>
    ''' Editiert den in UserInfo und AddressDetails gespeicherten Benutzer.
    ''' </summary>
    ''' <param name="ui">UserInfo mit den Hauptdaten für den Benutzer.</param>
    ''' <param name="addrDet">AddressDetailsInfo mit den Adressendaten für den Benutzer.</param>
    ''' <returns>Neue (bei Versionshandling) bzw. alte ID des Users.</returns>
    ''' <remarks>Die korrelierende Stored Procedure handelt die Versionskontrolle und schließt doppelte Benutzernamen aus.</remarks>
    Public Function Users_Edit(ByVal ui As UserInfo, ByVal LastEditedByIDUser As Integer, ByVal addrDet As AddressDetailsInfo) As Integer

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("Users_Edit", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = ui.IDSubsidiary
                .Add("@IDUser", SqlDbType.Int).Value = ui.IDUser
                .Add("@IDCostCenter", SqlDbType.Int).Value = ui.IDCostCenter
                .Add("@FirstName", SqlDbType.NVarChar, 100).Value = ui.FirstName
                .Add("@LastName", SqlDbType.NVarChar, 100).Value = ui.LastName
                .Add("@UserName", SqlDbType.NVarChar, 100).Value = ui.Username
                .Add("@Password", SqlDbType.VarBinary, 128).Value = ui.Password
                .Add("@ClearanceLevel", SqlDbType.BigInt).Value = ui.ClearanceLevel
                .Add("@HasWorkstationAccess", SqlDbType.Bit).Value = ui.HasWorkstationAccess
                .Add("@HasInternetAccess", SqlDbType.Bit).Value = ui.HasInternetAccess
                .Add("@IsActivated", SqlDbType.Bit).Value = ui.IsActivated
                .Add("@DoesExpire", SqlDbType.Bit).Value = ui.DoesExpire
                .Add("@ExpireDate", SqlDbType.DateTime).Value = ui.ExpireDate
                .Add("@Comment", SqlDbType.NText).Value = ui.Comment.Value
                .Add("@LastEditedByIDUser", SqlDbType.Int).Value = LastEditedByIDUser

                'Adressen-Details
                .Add("@PersonnelNo", SqlDbType.Int).Value = addrDet.PersonnelNo.Value
                .Add("@MiddleName", SqlDbType.NVarChar, 100).Value = addrDet.MiddleName.Value
                .Add("@Title", SqlDbType.NVarChar, 100).Value = addrDet.Titel.Value
                .Add("@Street", SqlDbType.NVarChar, 100).Value = addrDet.Street.Value
                .Add("@Zip", SqlDbType.NVarChar, 10).Value = addrDet.Zip.Value
                .Add("@City", SqlDbType.NVarChar, 100).Value = addrDet.City.Value
                .Add("@CountryCode", SqlDbType.NVarChar, 10).Value = addrDet.CountryCode.Value
                .Add("@Country", SqlDbType.NVarChar, 100).Value = addrDet.Country.Value
                .Add("@CompanyPhone", SqlDbType.NVarChar, 100).Value = addrDet.CompanyPhone.Value
                .Add("@PrivatePhone", SqlDbType.NVarChar, 100).Value = addrDet.PrivatePhone.Value
                .Add("@CompanyEmail", SqlDbType.NVarChar, 255).Value = addrDet.CompanyEmail.Value
                .Add("@PrivateEmail", SqlDbType.NVarChar, 255).Value = addrDet.PrivateEmail.Value
                .Add("@CompanyMobile", SqlDbType.NVarChar, 100).Value = addrDet.CompanyMobile.Value
                .Add("@PrivateMobile", SqlDbType.NVarChar, 100).Value = addrDet.PrivateMobile.Value
                .Add("@URL", SqlDbType.NVarChar, 100).Value = addrDet.URL.Value
                .Add("@ConsiderHistoryMaintenance", SqlDbType.Bit).Value = ConsiderHistoryMaintenance
                .Add("@IDUserNew", SqlDbType.Int)
                .Item("@IDUserNew").Direction = ParameterDirection.Output
            End With
            locCmd.ExecuteReader()
            Return CInt(locCmd.Parameters("@IDUserNew").Value)
        End Using
    End Function

    Public ReadOnly Property UserInfoCollection() As UserInfoCollection
        Get
            Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
            If locConnection Is Nothing Then
                Return Nothing
            End If
            Using locConnection
                Dim locCommand As New SqlCommand("SELECT * From [Users] WHERE [IDSubsidiary]=" & _
                    FacessoGeneric.LoginInfo.IDSubsidiary & " AND [IsSystemAccount]=0", locConnection)
                Dim locDR As SqlDataReader = locCommand.ExecuteReader()
                If locDR.HasRows Then
                    Dim locUIC As New UserInfoCollection
                    Do While locDR.Read
                        Dim locUserInfo As New UserInfo()
                        locUserInfo.AssignFieldsFromDataReader(locDR)
                        locUIC.Add(locUserInfo)
                    Loop
                    Return locUIC
                Else
                    Return Nothing
                End If
            End Using
        End Get
    End Property

End Class

File: src/MixNetFx/Facesso.Data/SQLServer/SPAccess_Users.cs
using System;
using System.Data;
using System.Data.SqlClient;
using ActiveDev;
using Facesso;

namespace Facesso.Data
{
    public sealed partial class SPAccess
    {
        public bool Users_DoesUsernameExist(int idSubsidiary, string username, ADDBNullable<int> excludeIDUser)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return false;
            using (locConnection)
            {
                var locCmd = new SqlCommand("Users_DoesUsernameExist", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = idSubsidiary;
                locCmd.Parameters.Add("@Username", SqlDbType.NVarChar, 100).Value = username;
                locCmd.Parameters.Add("@ExcludeIDUser", SqlDbType.Int).Value = excludeIDUser.Value;
                locCmd.Parameters.Add("@DoesExist", SqlDbType.Bit);
                locCmd.Parameters["@DoesExist"].Direction = ParameterDirection.Output;
                locCmd.ExecuteReader();
                return (bool)locCmd.Parameters["@DoesExist"].Value;
            }
        }

        public int Users_Add(UserInfo ui, int createdByIDUser, AddressDetailsInfo addrDet)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return 0;
            using (locConnection)
            {
                var locCmd = new SqlCommand("Users_Add", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = ui.IDSubsidiary;
                locCmd.Parameters.Add("@IDCostCenter", SqlDbType.Int).Value = ui.IDCostCenter;
                locCmd.Parameters.Add("@FirstName", SqlDbType.NVarChar, 100).Value = ui.FirstName;
                locCmd.Parameters.Add("@LastName", SqlDbType.NVarChar, 100).Value = ui.LastName;
                locCmd.Parameters.Add("@UserName", SqlDbType.NVarChar, 100).Value = ui.Username;
                locCmd.Parameters.Add("@Password", SqlDbType.VarBinary, 128).Value = ui.Password;
                locCmd.Parameters.Add("@ClearanceLevel", SqlDbType.BigInt).Value = ui.ClearanceLevel;
                locCmd.Parameters.Add("@HasWorkstationAccess", SqlDbType.Bit).Value = ui.HasWorkstationAccess;
                locCmd.Parameters.Add("@HasInternetAccess", SqlDbType.Bit).Value = ui.HasInternetAccess;
                locCmd.Parameters.Add("@IsActivated", SqlDbType.Bit).Value = ui.IsActivated;
                locCmd.Parameters.Add("@DoesExpire", SqlDbType.Bit).Value = ui.DoesExpire;
                locCmd.Parameters.Add("@ExpireDate", SqlDbType.DateTime).Value = ui.ExpireDate;
                locCmd.Parameters.Add("@WasCurrentTo", SqlDbType.DateTime).Value = FacessoGeneric.OpenCurrentToDate;
                locCmd.Parameters.Add("@CreatedByIDUser", SqlDbType.Int).Value = createdByIDUser;
                locCmd.Parameters.Add("@Comment", SqlDbType.NText).Value = ui.Comment.Value;
                // Address details
                locCmd.Parameters.Add("@PersonnelNo", SqlDbType.Int).Value = addrDet.PersonnelNo.Value;
                locCmd.Parameters.Add("@MiddleName", SqlDbType.NVarChar, 100).Value = addrDet.MiddleName.Value;
                locCmd.Parameters.Add("@Title", SqlDbType.NVarChar, 100).Value = addrDet.Titel.Value;
                locCmd.Parameters.Add("@Street", SqlDbType.NVarChar, 100).Value = addrDet.Street.Value;
                locCmd.Parameters.Add("@Zip", SqlDbType.NVarChar, 10).Value = addrDet.Zip.Value;
                locCmd.Parameters.Add("@City", SqlDbType.NVarChar, 100).Value = addrDet.City.Value;
                locCmd.Parameters.Add("@CountryCode", SqlDbType.NVarChar, 10).Value = addrDet.CountryCode.Value;
                locCmd.Parameters.Add("@Country", SqlDbType.NVarChar, 100).Value = addrDet.Country.Value;
                locCmd.Parameters.Add("@CompanyPhone", SqlDbType.NVarChar, 100).Value = addrDet.CompanyPhone.Value;
                locCmd.Parameters.Add("@PrivatePhone", SqlDbType.NVarChar, 100).Value = addrDet.PrivatePhone.Value;
                locCmd.Parameters.Add("@CompanyEmail", SqlDbType.NVarChar, 255).Value = addrDet.CompanyEmail.Value;
                locCmd.Parameters.Add("@PrivateEmail", SqlDbType.NVarChar, 255).Value = addrDet.PrivateEmail.Value;
                locCmd.Parameters.Add("@CompanyMobile", SqlDbType.NVarChar, 100).Value = addrDet.CompanyMobile.Value;
                locCmd.Parameters.Add("@PrivateMobile", SqlDbType.NVarChar, 100).Value = addrDet.PrivateMobile.Value;
                locCmd.Parameters.Add("@URL", SqlDbType.NVarChar, 100).Value = addrDet.URL.Value;
                locCmd.Parameters.Add("@IDUserNew", SqlDbType.Int);
                locCmd.Parameters["@IDUserNew"].Direction = ParameterDirection.Output;
                locCmd.ExecuteReader();
                return (int)locCmd.Parameters["@IDUserNew"].Value;
            }
        }

        public int Users_Edit(UserInfo ui, int lastEditedByIDUser, AddressDetailsInfo addrDet)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return 0;
            using (locConnection)
            {
                var locCmd = new SqlCommand("Users_Edit", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = ui.IDSubsidiary;
                locCmd.Parameters.Add("@IDUser", SqlDbType.Int).Value = ui.IDUser;
                locCmd.Parameters.Add("@IDCostCenter", SqlDbType.Int).Value = ui.IDCostCenter;
                locCmd.Parameters.Add("@FirstName", SqlDbType.NVarChar, 100).Value = ui.FirstName;
                locCmd.Parameters.Add("@LastName", SqlDbType.NVarChar, 100).Value = ui.LastName;
                locCmd.Parameters.Add("@UserName", SqlDbType.NVarChar, 100).Value = ui.Username;
                locCmd.Parameters.Add("@Password", SqlDbType.VarBinary, 128).Value = ui.Password;
                locCmd.Parameters.Add("@ClearanceLevel", SqlDbType.BigInt).Value = ui.ClearanceLevel;
                locCmd.Parameters.Add("@HasWorkstationAccess", SqlDbType.Bit).Value = ui.HasWorkstationAccess;
                locCmd.Parameters.Add("@HasInternetAccess", SqlDbType.Bit).Value = ui.HasInternetAccess;
                locCmd.Parameters.Add("@IsActivated", SqlDbType.Bit).Value = ui.IsActivated;
                locCmd.Parameters.Add("@DoesExpire", SqlDbType.Bit).Value = ui.DoesExpire;
                locCmd.Parameters.Add("@ExpireDate", SqlDbType.DateTime).Value = ui.ExpireDate;
                locCmd.Parameters.Add("@Comment", SqlDbType.NText).Value = ui.Comment.Value;
                locCmd.Parameters.Add("@LastEditedByIDUser", SqlDbType.Int).Value = lastEditedByIDUser;
                // Address details
                locCmd.Parameters.Add("@PersonnelNo", SqlDbType.Int).Value = addrDet.PersonnelNo.Value;
                locCmd.Parameters.Add("@MiddleName", SqlDbType.NVarChar, 100).Value = addrDet.MiddleName.Value;
                locCmd.Parameters.Add("@Title", SqlDbType.NVarChar, 100).Value = addrDet.Titel.Value;
                locCmd.Parameters.Add("@Street", SqlDbType.NVarChar, 100).Value = addrDet.Street.Value;
                locCmd.Parameters.Add("@Zip", SqlDbType.NVarChar, 10).Value = addrDet.Zip.Value;
                locCmd.Parameters.Add("@City", SqlDbType.NVarChar, 100).Value = addrDet.City.Value;
                locCmd.Parameters.Add("@CountryCode", SqlDbType.NVarChar, 10).Value = addrDet.CountryCode.Value;
                locCmd.Parameters.Add("@Country", SqlDbType.NVarChar, 100).Value = addrDet.Country.Value;
                locCmd.Parameters.Add("@CompanyPhone", SqlDbType.NVarChar, 100).Value = addrDet.CompanyPhone.Value;
                locCmd.Parameters.Add("@PrivatePhone", SqlDbType.NVarChar, 100).Value = addrDet.PrivatePhone.Value;
                locCmd.Parameters.Add("@CompanyEmail", SqlDbType.NVarChar, 255).Value = addrDet.CompanyEmail.Value;
                locCmd.Parameters.Add("@PrivateEmail", SqlDbType.NVarChar, 255).Value = addrDet.PrivateEmail.Value;
                locCmd.Parameters.Add("@CompanyMobile", SqlDbType.NVarChar, 100).Value = addrDet.CompanyMobile.Value;
                locCmd.Parameters.Add("@PrivateMobile", SqlDbType.NVarChar, 100).Value = addrDet.PrivateMobile.Value;
                locCmd.Parameters.Add("@URL", SqlDbType.NVarChar, 100).Value = addrDet.URL.Value;
                locCmd.Parameters.Add("@ConsiderHistoryMaintenance", SqlDbType.Bit).Value = FacessoGeneric.ConsiderHistoryMaintenance;
                locCmd.Parameters.Add("@IDUserNew", SqlDbType.Int);
                locCmd.Parameters["@IDUserNew"].Direction = ParameterDirection.Output;
                locCmd.ExecuteReader();
                return (int)locCmd.Parameters["@IDUserNew"].Value;
            }
        }

        public UserInfoCollection UserInfoCollection
        {
            get
            {
                SqlConnection locConnection = GetOpenedConnectionSafely();
                if (locConnection == null) return null;
                using (locConnection)
                {
                    var locCommand = new SqlCommand(
                        "SELECT * From [Users] WHERE [IDSubsidiary]=@IDSubsidiary AND [IsSystemAccount]=0",
                        locConnection);
                    locCommand.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = FacessoGeneric.LoginInfo.IDSubsidiary;
                    SqlDataReader locDR = locCommand.ExecuteReader();
                    if (locDR.HasRows)
                    {
                        var locUIC = new UserInfoCollection();
                        while (locDR.Read())
                        {
                            var locUserInfo = new UserInfo();
                            locUserInfo.AssignFieldsFromDataReader(locDR);
                            locUIC.Add(locUserInfo);
                        }
                        return locUIC;
                    }
                    return null;
                }
            }
        }
    }
}
